### PR TITLE
Add on-progress-update-fn callback

### LINKED ISSUE #1675: Support callback function for each file analyzed
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
As mentioned [here](https://clojurians.slack.com/archives/CHY97NXE2/p1651283825804139), it'd be super helpful if kondo could provide a callback function that would be called for each analyzed filename, so clojure-lsp could update progress bar smarter and precisely.

**Describe the solution you'd like**
A custom callback function on clj-kondo JVM which would be called N times being N the number of filenames analyzed.
Note: if passing `src`, it'd be even better if kondo could call the callback for each file not only one time.

**Describe alternatives you've considered**

**Additional context**

I can open a PR, any suggestion on where to add that? maybe near existing custom-lint-fn, not sure all files are already available there to loop

### PR DESCRIPTION
**Pull request by ericdallo**
Fixes #1675 

- [X] I have read the [developer documentation](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md).

- [X] This PR correponds to an [issue with a clear problem statement](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).

- [X] This PR contains a [test](https://github.com/clj-kondo/clj-kondo/blob/master/doc/dev.md#tests) to prevent against future regressions

- [X] I have updated the [CHANGELOG.md](https://github.com/clj-kondo/clj-kondo/blob/master/CHANGELOG.md) file with a description of the addressed issue.

### CODE REVIEW

File: src/clj_kondo/impl/core.clj [hunk 1]
```diff
@@ -408,21 +411,26 @@
...
                          :lang (if filename
                                  (lang-from-file filename default-language)
                                  default-language)} dev?)
           (classpath? path)
-          (run! #(process-file ctx % default-language canonical? filename)
-                (str/split path
-                           (re-pattern path-separator)))
+          (let [paths (str/split path
+                                 (re-pattern path-separator))]
+            (run! #(process-file ctx % default-language canonical? filename (count paths))
```
Comment (borkdude) [inline src/clj_kondo/impl/core.clj:432]: It seems this count is supposed to be total-files, but it's actually the number of classpath entries here, so jars + directories.

Comment (ericdallo) [reply src/clj_kondo/impl/core.clj:432]: yeah, indeed, any suggestion how to get all possible files here?

**Comment by ericdallo** [comment]
Not sure why 3 tests are failing, still debugging it

File: src/clj_kondo/impl/core.clj (hunk 1 above)
Comment (ericdallo) [reply src/clj_kondo/impl/core.clj:432]: Also, I'm ok on we don't have all specific files, but something roughly, a different name would be better though if that was the case

Comment (borkdude) [reply src/clj_kondo/impl/core.clj:432]: In parallel mode you don't know the number of files before hand. So I don't think we can reliably add this number at all.

Comment (ericdallo) [reply src/clj_kondo/impl/core.clj:432]: RIght, I think I can try to make that on lsp side, probably will be easier and enough for what we want to fix: a better feedback, it doesn't need to be totally precise. I let you know if that works and then change this PR

**Comment by ericdallo** [comment]
@borkdude It took me some time but it works as expected now, I did some changes and spent some time making the parallel work, also, I made the feature a little more specific, now it's called for each entry passed on :lint, not for every file as we can't know when :parallel as you said, this is enough for clojure-lsp have a way better progress report. LMK what you think

**Comment by borkdude** [comment]
@ericdallo Can you explain in words to me how it now works and how this benefits clojure-lsp and potentially other projects?

**Comment by ericdallo** [comment]
Sure, the idea is to have a callback function related to the progress of the analysis, the first idea was to call that for each filename but since this is not possible to know the total times that function would be called, it would make hard to clojure-lsp or any client create some kind of progress bar without knowing how many times the function would be called.

The way it works now is: it calls that callback for each entry passed which could be a file or a folder, check the tests where those cases are tested. The parallel mode made a little bit more trick to make it work, but in the end the logic is the same

**Comment by ericdallo** [comment]
Alright @borkdude, made one more change, now code is way easier since we count total files before when that callback is passed, and the callback is `file-analyzed-fn` which will be triggered for every filename analyzed

**Comment by borkdude** [comment]
I'll look again when the tests are fixed :)

**Comment by ericdallo** [comment]
Ops, my bad, fixed @borkdude !

File: src/clj_kondo/impl/core.clj [hunk 2]
```diff
@@ -361,6 +361,47 @@
 (defn classpath? [f]
   (str/includes? f path-separator))
 
+(defn ^:private entries-from-jar [^java.io.File jar-file]
```
Comment (borkdude) [inline src/clj_kondo/impl/core.clj:364]: I think laziness should never escape `with-open`. Also, perhaps we can just call this `entries-from-jar-count`, and already count in there before closing?

File: src/clj_kondo/impl/core.clj [hunk 3]
```diff
@@ -361,6 +361,47 @@
...
+              (let [path (str path)
+                    file (io/file path)
+                    canonical (when (.exists file)
+                                (.getCanonicalPath file))]
+                (cond
+                  (and canonical
+                       (str/ends-with? canonical ".jar"))
+                  (count (entries-from-jar file))
+
+                  (and canonical
```
Comment (borkdude) [inline src/clj_kondo/impl/core.clj:384]: The canonical stuff doesn't really matter for counting files I think?

Comment (ericdallo) [reply src/clj_kondo/impl/core.clj:384]: yes, this is just the path, maybe we could rename from canonical to other name?

Comment (borkdude) [reply src/clj_kondo/impl/core.clj:384]: ok

**Comment by borkdude** [comment]
Also Windows tests are failing.

**Comment by borkdude** [comment]
There is some normalize-path function in the test-utils to fix that.

File: src/clj_kondo/impl/analyzer.clj [hunk 4]
```diff
@@ -2707,7 +2707,7 @@
             (print ".") (flush))))
       (when file-analyzed-fn
         (file-analyzed-fn {:filename filename
```
[the remark this answers is not in the document]
Comment (ericdallo) [reply src/clj_kondo/impl/analyzer.clj:2709]: I think we can assoc both for now?

[the remark this answers is not in the document]
Comment (borkdude) [reply src/clj_kondo/impl/analyzer.clj:2709]: OK

**Pull request merged**